go/ql/lib/semmle/go/frameworks/Twirp.qll

```diff
@@ -0,0 +1,174 @@
+
+    /**
+     * Returns the named type on top of this interface type
+     */
+    NamedType getNamedType() { result = namedType }
+  }
+
+  /**
+   * A Twirp client.
+   */
+  class ServiceClientType extends NamedType {
+    ServiceClientType() {
+      exists(ServiceInterfaceType i, PointerType p |
+        p.implements(i) and
+        this = p.getBaseType() and
+        this.getName().toLowerCase() = i.getName().toLowerCase() + ["protobuf", "json"] + "client"
+      )
+    }
+  }
+
+  /**
+   * A Twirp server
+   */
+  class ServiceServerType extends NamedType {
+    ServiceServerType() {
+      exists(ServiceInterfaceType i |
+        this.implements(i) and
+        this.getName().toLowerCase() = i.getName().toLowerCase() + "server"
+      )
+    }
+  }
+
+  /**
+   * A Twirp function to construct a Client
+   */
+  class ClientConstructor extends Function {
+    ClientConstructor() {
+      exists(ServiceClientType c |
+        this.getName().toLowerCase() = "new" + c.getName().toLowerCase() and
+        this.getParameter(0).getType() instanceof StringType and
+        this.getParameterType(1).getName() = "HTTPClient"
+      )
+    }
+  }
+
+  /**
+   * A Twirp function to construct a Server
+   * Its first argument should be an implementation of the service interface
+   */
+  class ServerConstructor extends Function {
+    ServerConstructor() {
+      exists(ServiceServerType c, ServiceInterfaceType i |
+        this.getName().toLowerCase() = "new" + c.getName().toLowerCase() and
+        this.getParameter(0).getType() = i.getNamedType()
+      )
+    }
+  }
+
+  /**
+   * An SSRF sink for the Client constructor
+   */
+  class ClientRequestUrlAsSink extends RequestForgery::Sink {
+    ClientRequestUrlAsSink() {
+      exists(DataFlow::CallNode call |
+        call.getArgument(0) = this and
+        call.getTarget() instanceof ClientConstructor
+      )
+    }
+
+    override DataFlow::Node getARequest() { none() }
+
+    override string getKind() { result = "URL" }
+  }
+
+  /**
+   * A service handler
+   */
+  class ServiceHandler extends Method {
+    ServiceHandler() {
+      exists(DataFlow::CallNode call, Type handlerType, ServiceInterfaceType i |
+        call.getTarget() instanceof ServerConstructor and
+        call.getArgument(0).getType() = handlerType and
+        this = handlerType.getMethod(_) and
+        this.implements(i.getNamedType().getMethod(_))
+      )
+    }
+  }
+
+  /**
+   * A request coming to the service handler
+   */
+  class Request extends UntrustedFlowSource::Range, DataFlow::ParameterNode {
+    ServiceHandler handler;
+
+    Request() {
+      handler.getParameter(0).getType().hasQualifiedName("context", "Context") and
+      handler.getParameter(_) = this.asParameter() and
+      this.getType().(PointerType).getBaseType() instanceof ProtobufMessageType
+    }
+
+    override predicate isParameterOf(Callable c, int i) {
+      c.asFunction() = handler and
+      i = 1
+    }
+  }
```

Comment:
```suggestion

  /**
   * A request coming to the service handler.
   */
  class Request extends UntrustedFlowSource::Range instanceof DataFlow::ParameterNode {
    Request() {
      exists(Callable c, ServiceHandler handler | c.asFunction() = handler |
        this.isParameterOf(c, 1) and
        handler.getParameterType(0).hasQualifiedName("context", "Context") and
        this.getType().(PointerType).getBaseType() instanceof ProtobufMessageType
      )
    }
  }
```
By `extend`ing an abstract class you are actually making new parameter nodes. Better to use `instanceof` instead, so you're only using ones that already exist.

Reply:
Thanks, good to know!